README.md
# HTTP Request Logger · Eliasis PHP Framework plugin

[![Packagist](https://img.shields.io/packagist/v/eliasis-framework/http-request-logger.svg)](https://packagist.org/packages/eliasis-framework/http-request-logger)
[![License](https://img.shields.io/packagist/l/eliasis-framework/http-request-logger.svg)](https://github.com/eliasis-framework/http-request-logger/blob/master/LICENSE)

[Versión en español](README-ES.md)

Save HTTP request information to the database.

---

- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
- [Tests](#tests)
- [Patrocinar](#patrocinar)
- [Licencia](#licencia)

---

## Requirements

This plugin is supported by **PHP versions 5.6** or higher and is compatible with **HHVM versions 3.0** or higher.

## Installation

The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **HTTP Request Logger**, simply:

    composer require eliasis-framework/http-request-logger

The previous command will only install the necessary files, if you prefer to **download the entire source code** you can use:

    composer require eliasis-framework/http-request-logger --prefer-source

You can also **clone the complete repository** with Git:

    git clone https://github.com/eliasis-framework/http-request-logger.git

## Usage

To use this plugin, your [Eliasis application](https://github.com/eliasis-framework/eliasis) must use the [PHP-Database](https://eliasis-framework.github.io/eliasis/v1.1.3/lang/en/#libraries-Database) library and add the following to the application configuration files:

```php
/**
 * eliasis-app/config/complements.php
 */
return [

    'plugin' => [

        'http-request-logger' => [

            'db-id' => 'app',
            'db-prefix' => 'test_',
            'db-charset' => 'utf8',
            'db-engine' => 'innodb'
        ],
    ],
];
```

This will create the `test_request` table and automatically save all HTTP requests.

The table structure created is as follows:

| Columns | Data type | Example |
| --- | --- | --- |
| request_id | INT(9) | 1 |
| request_ip | VARCHAR(45) | 87.142.85.70 |
| request_uri | TEXT | /sample-app/ |
| request_protocol | VARCHAR(100) | HTTP/1.1 |
| request_method | VARCHAR(15) | GET |
| request_referer | VARCHAR(255) | <http://www.google.es/> |
| request_user_agent | VARCHAR(255) | Mozilla/5.0 (...) |
| request_http_state | INT(3) | 200 |
| request_load_time | FLOAT | 0.008 |
| created | TIMESTAMP | 2018-02-28 08:26:43 |

## Tests

To run [tests](tests) you just need [composer](http://getcomposer.org/download/) and to execute the following:

    git clone https://github.com/eliasis-framework/http-request-logger.git
    
    cd http-request-logger

    composer install

Run unit tests with [PHPUnit](https://phpunit.de/):

    composer phpunit

Run [PSR2](http://www.php-fig.org/psr/psr-2/) code standard tests with [PHPCS](https://github.com/squizlabs/PHP_CodeSniffer):

    composer phpcs

Run [PHP Mess Detector](https://phpmd.org/) tests to detect inconsistencies in code style:

    composer phpmd

Run all previous tests:

    composer tests

## Patrocinar

Si este proyecto te ayuda a reducir el tiempo de desarrollo,
[puedes patrocinarme](https://github.com/josantonius/lang/es-ES/README.md#patrocinar)
para apoyar mi trabajo :blush:

## Licencia

Este repositorio tiene una licencia [MIT License](LICENSE).

Copyright © 2017-2022, [Josantonius](https://github.com/josantonius/lang/es-ES/README.md#contacto)


CHANGELOG.md
# CHANGELOG

## 2022-08-19

* The repository was archived.

## 1.0.1 - 2018-02-28

* Implemented `PHP Mess Detector` to detect inconsistencies in code styles.

* Implemented `PHP Code Beautifier and Fixer` to fixing errors automatically.

* Implemented `PHP Coding Standards Fixer` to organize PHP code automatically according to PSR standards.

* Implemented `PSR-4 autoloader standard` from all library files.

* Implemented `PSR-2 coding standard` from all library PHP files.

* Implemented `PHPCS` to ensure that PHP code complies with `PSR2` code standards.

* Implemented `Codacy` to automates code reviews and monitors code quality over time.

* Implemented `Codecov` to coverage reports.

## 1.0.0 - 2017-07-02

* Added `App\Modules\Request\Controller\Exception\RequestException` class.
* Added `App\Modules\Request\Controller\Exception\RequestException->__construct()` method.

* Added `App\Modules\Request\Controller\Launcher\Launcher` class.
* Added `App\Modules\Request\Controller\Launcher\Launcher->init()` method.
* Added `App\Modules\Request\Controller\Launcher\Launcher->getDatabaseId()` method.
* Added `App\Modules\Request\Controller\Launcher\Launcher->activation()` method.
* Added `App\Modules\Request\Controller\Launcher\Launcher->deactivation()` method.

* Added `App\Modules\Request\Controller\Request\Request` class.
* Added `App\Modules\Request\Controller\Request\Request->createTable()` method.
* Added `App\Modules\Request\Controller\Request\Request->dropTable()` method.
* Added `App\Modules\Request\Controller\Request\Request->insert()` method.

* Added `App\Modules\Request\Model\Request\Request` class.
* Added `App\Modules\Request\Model\Request\Request->__construct()` method.
* Added `App\Modules\Request\Model\Request\Request->createTable()` method.
* Added `App\Modules\Request\Model\Request\Request->dropTable()` method.
* Added `App\Modules\Request\Model\Request\Request->insert()` method.

* Added `config/namespaces.php` file.
* Added `config/server.php` file.
* Added `config/set-hooks.php` file.
* Added `config/tables.php` file.

* Added `request.php` file.


src/Model/Request/Request.php
<?php
/**
 * Eliasis PHP Framework Request module
 * 
 * @author     Josantonius - hello@josantonius.com
 * @copyright  Copyright (c) 2017
 * @license    https://opensource.org/licenses/MIT - The MIT License (MIT)
 * @link       https://github.com/Eliasis-Framework/Request.git
 * @since      1.0.0
 */

namespace Eliasis\Modules\Request\Model\Request;

use Eliasis\App\App,
    Eliasis\Model\Model,
    Eliasis\Module\Module,
    Josantonius\Database\Database;
    
/**
 * Module main model.
 *
 * @since 1.0.0
 */
class Request extends Model {

    /**
     * Database table name.
     *
     * @since 1.0.0
     *
     * @var string $id
     */
    public $id = 'request';

    /**
     * Database object connection.
     *
     * @since 1.0.0
     *
     * @var object $db
     */
    public $db;

    /**
     * Database table columns.
     *
     * @since 1.0.0
     *
     * @var array $data
     */
    public $data;

    /**
     * Get database connection.
     *
     * @since 1.0.0
     */
    public function __construct() {

        $database = Module::Request()->get('db-id');
        
        $this->db = Database::getConnection($database);

        $this->data = Module::Request()->get('table', $this->id);
    }

    /**
     * Create table.
     * 
     * @return boolean
     */
    public function createTable() {

        $params = [
            $this->data['id'] => 'INT(9) AUTO_INCREMENT PRIMARY KEY', 
            $this->data['ip']         => 'VARCHAR(15)  NOT NULL',
            $this->data['uri']        => 'TEXT         NOT NULL',
            $this->data['protocol']   => 'VARCHAR(100) NOT NULL',
            $this->data['method']     => 'VARCHAR(15)  NOT NULL',
            $this->data['referer']    => 'VARCHAR(255) NOT NULL',
            $this->data['user_agent'] => 'VARCHAR(255) NOT NULL',
            $this->data['http_state'] => 'INT(3)       NOT NULL',
            $this->data['resp_state'] => 'INT(3)       NOT NULL',
            $this->data['load_time']  => 'FLOAT        NOT NULL',
            $this->data['created']    => 'TIMESTAMP DEFAULT ' .
                                         'CURRENT_TIMESTAMP',
        ];
            
        $query = $this->db->create($params)
                          ->table($this->data['prefix'] . $this->id)
                          ->engine($this->data['engine'])
                          ->charset($this->data['charset']);

        return $query->execute();
    }

    /**
     * Insert row.
     *
     * @since 1.0.0
     *
     * @param string $ip        → request ip
     * @param string $uri       → request uri
     * @param string $protocol  → request protocol
     * @param string $method    → request method
     * @param string $referer   → request referer
     * @param string $userAgent → request user agent
     * @param int    $httpState → request http state
     * @param int    $respState → request response state
     * @param float  $loadTime  → request load time
     *
     * @return int → id inserted
     */
    public function insert($ip, $uri, $protocol, $method, $referer, $userAgent, $httpState, $respState, $loadTime) {

        $data = [

            $this->data['ip']         => '?',
            $this->data['uri']        => '?',
            $this->data['protocol']   => '?', 
            $this->data['method']     => '?',
            $this->data['referer']    => '?',
            $this->data['user_agent'] => '?',
            $this->data['http_state'] => '?',
            $this->data['resp_state'] => '?',
            $this->data['load_time']  => '?',
        ];

        $statements[] = [1, $ip,        'str'];
        $statements[] = [2, $uri,       'str'];
        $statements[] = [3, $protocol,  'str'];
        $statements[] = [4, $method,    'str'];
        $statements[] = [5, $referer,   'str'];
        $statements[] = [6, $userAgent, 'str'];
        $statements[] = [7, $httpState, 'int'];
        $statements[] = [8, $respState, 'int'];
        $statements[] = [9, $loadTime];

        $query = $this->db->insert($data, $statements)
                          ->from($this->data['prefix'] . $this->id);

        return $query->execute('id');
    }

    /**
     * Delete requests table.
     *
     * @since 1.0.0
     *
     * @return boolean
     */
    public function dropTable() {

        $query = $this->db->drop()
                          ->table($this->data['prefix'] . $this->id);

        return $query->execute();
    }
}
